src/opt/horizontal_fusion_impl.cpp

```diff
@@ -0,0 +1,839 @@
+        {
+            if(ndx > axis)
+                unit_slice *= len;
+            if(ndx != axis)
+                new_elements *= len;
+            ndx++;
+        }
+        new_elements *= sum;
+        unsigned type_size             = s.type_size();
+        unsigned long long total_bytes = new_elements * type_size;
+        std::shared_ptr<char> input    = make_shared_array<char>(total_bytes);
+        std::vector<unsigned long long> bytes_per_slice;
+        unsigned long long unit_slice_bytes = unit_slice * type_size;
+
+        std::transform(instrs.begin(),
+                       instrs.end(),
+                       std::back_inserter(bytes_per_slice),
+                       [&](auto&& d) -> unsigned long long {
+                           return d->get_shape().lens().at(axis) * unit_slice_bytes;
+                       });
+
+        unsigned copy_bytes = 0;
+        int slice_ndx       = 0;
+        char* dst           = input.get();
+        while(copy_bytes < total_bytes)
+        {
+            unsigned ins_ndx = 0;
+            for(auto&& ins : instrs)
+            {
+                unsigned long long bytes = bytes_per_slice[ins_ndx];
+                const char* src          = &(ins->get_literal().data()[slice_ndx * bytes]);
+                std::copy(src, src + bytes, dst);
+                dst += bytes;
+                copy_bytes += bytes;
+                ins_ndx++;
+            }
+            slice_ndx++;
+        }
+        shape new_shape{s.type(), new_lens};
+        auto new_literal = p_program->add_literal(literal(new_shape, input.get()));
+        instruction::replace_argument(output, ins0, new_literal, false);
+    }
+    else
+    {
+        ins0->set_shape({ins0->get_shape().type(), new_lens});
+        if(ins0->name() == "broadcast")
+        {
+            // workaround for a bad practice: broadcast has a broadcast_shape field.
+            uint64_t ax  = (any_cast<op::broadcast>(ins0->get_operator())).axis;
+            operation op = op::broadcast{ax, ins0->get_shape().lens()};
+            std::vector<shape> input_shapes;
+            input_shapes.push_back(ins0->inputs().at(0)->get_shape());
+            shape new_s = op.compute_shape(input_shapes);
+            ins0->set_operator(op);
+            ins0->set_shape(new_s);
+        }
+    }
+
+    if(output == base)
+        output->set_shape({output->get_shape().type(), base_lens});
+}
+
+// If ins and input only diff in one axis, return that axis.
+int horizontal_fusion_impl::find_unique_axis(instruction_ref ins, instruction_ref input)
+{
+    auto lens1 = ins->get_shape().lens();
+    auto lens2 = input->get_shape().lens();
+    if(lens1.size() != lens2.size())
+        return -1;
+    int count = 0;
+    int ndx   = 0;
+    int ret   = -1;
+    for(auto&& size : lens1)
+    {
+        if(size != lens2.at(ndx))
+        {
+            count++;
+            ret = ndx;
+        }
+        ndx++;
+    }
+    return (count == 1) ? ret : -1;
```

Comment:
This can be written with `std::mismatch` and `std::equal`:

```cpp
int horizontal_fusion_impl::find_unique_axis(instruction_ref ins, instruction_ref input)
{
    auto lens1 = ins->get_shape().lens();
    auto lens2 = input->get_shape().lens();
    auto m = std::mismatch(lens1.begin(), lens1.end(), lens2.begin(), lens2.end());
    if ((lens1.end() - m.first > 1 or lens2.end() - m.second > 1) and 
        not std::equal(m.first+1, lens1.end(), m.second+1, lens2.end()))
        return -1;
    return m.first - lens1.begin();
}
```

Reply:
fixed.